enhancements/insights/insights-runtime-extractor.md

```diff
@@ -0,0 +1,652 @@
+
+#### Performance Risk & Mitigation
+
+##### Limited Restrictions for Fingerprints
+
+Since a fingerprint executable runs under a downgraded privilege model, with a limited process life-span, significantly more freedom is afforded to it, but some care is still necessary. Notably, a fingerprint executable can have a larger dependency set beyond the core language SDK to meet its needs. Additionally, fingerprint executables can leverage languages with a greater emphasis on productivity (e.g. Golang), provided the implementations still utilize input and output hardening techniques, and do not load potentially rogue code in the container. Each fingerprint implementation may even have a different language runtime than its peers. The following rules should be followed:
+
+1. __Inspecting code must be statically compiled and not rely on dynamic library loading.__ This is essentially already required to achieve stable execution on arbitrary container layers, including those with differing libc implementations, or even scratch containers with no additional libraries
+2. __All analysis I/O operations should be limit-bound.__ A container either through error or design may have files of arbitrary size, or files in a corrupt state. Any potentially unbounded operation (e.g. reading to the end of a file or following a reference data structure to its arbitrary end) is prohibited.
+3. __Memory and CPU usage should not be input derived.__ In addition to achieving system stability and performance goals, ensuring the overhead of a inspecting process does not increase significantly relative to input data is critical in preventing potential Denial-of-Service attacks relying on a specially crafted container filesystem.  
+
+##### No Impact on Workloads
+
+Inspecting containers must not disrupt the containers that are inspected and have an impact on their quality of service.
+The insights-runtime-extractor is running with its own memory and CPU requirements and does not consume resources from the inspected container.
+
+If the worker node is near limits, a burst of activity from the insights runtime extractor could lead to node instability. To prevent this,
+the insights runtime extractor will extract runtime information from processes one at a time and will not parallelize these tasks.
+Increasing the gathering time is a trade-off to prevent node instability.
+
+###### Risk of misbehaving code execution
+
+A few fingerprints relies on executing process in the user namespace (for example, executing `node --version` to determine the version of Node.js).
+There is a risk of side-effects if this execution does not match the expectation (if the `node` process is not the Node.js executable and its `--version` parameter triggers a `rm -rf /`).
+Executing such code should be done conservatively on well-identified workloads (`python`, `node`, etc.) if they provide a way to get their versions without side-effect.
```

Comment:
Another issue here is that when we exec processes inside a container, we are using up the container's resources. If it is already near its memory limit then it could get OOMKilled by the kernel.

Reply:
I see that we mention joining namespaces of the process. Let's clarify that we won't join the cgroups.

Reply:
I mention that we don't enter the `cgroups` in https://github.com/openshift/enhancements/blob/e083208a087afcc9745bd0c0e154174985e16c76/enhancements/insights/insights-runtime-extractor.md#insights-runtime-extractor-constraints.
Does that alleviate the issue? We are consuming the extractor's container resources and would not put any risk of OOME to the process container, right?